File: backend/src/main/java/cn/etarch/mao/session/service/TaskTerminalService.java
package cn.etarch.mao.session.service;

import cn.etarch.mao.notification.task.entity.TaskNotificationDelivery;
import cn.etarch.mao.notification.task.service.TaskNotificationDeliveryService;
import cn.etarch.mao.session.entity.Session;
import cn.etarch.mao.session.ws.StreamingWsRegistry;
import cn.etarch.mao.session.ws.WsEvent;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.Executor;

@Slf4j
@Service
public class TaskTerminalService {
    private final SessionService sessionService;
    private final StreamingWsRegistry registry;
    private final TaskNotificationDeliveryService deliveryService;
    private final Executor notificationExecutor;

    public TaskTerminalService(SessionService sessionService,
                               StreamingWsRegistry registry,
                               TaskNotificationDeliveryService deliveryService,
                               @Qualifier("taskNotificationExecutor") Executor notificationExecutor) {
        this.sessionService = sessionService;
        this.registry = registry;
        this.deliveryService = deliveryService;
        this.notificationExecutor = notificationExecutor;
    }

    public void finishExecution(Long sessionId, Long userId, String phase, String executionId) {
        if (!("COMPLETED".equals(phase) || "FAILED".equals(phase) || "CANCELLED".equals(phase))) {
            throw new IllegalArgumentException("Unsupported terminal phase: " + phase);
        }
        Session previous = sessionService.getSession(sessionId);
        // Once a session reaches a terminal phase it must not transition to another
        // terminal phase. This prevents a late cancel from overriding a COMPLETED/FAILED
        // session (and vice versa) when the request races the execution thread's finish.
        if (isTerminalPhase(previous.getPhase())) {
            log.info("Ignoring terminal transition for already-terminal session: sessionId={}, from={}, to={}",
                    sessionId, previous.getPhase(), phase);
            return;
        }
        sessionService.updatePhase(sessionId, phase);
        Session session = sessionService.getSession(sessionId);
        Long ownerId = userId != null ? userId : session.getUserId();

        Map<String, Object> statusData = new HashMap<>();
        statusData.put("phase", phase);
        statusData.put("unread", true);
        if (executionId != null && !executionId.isBlank()) statusData.put("executionId", executionId);

        Optional<TaskNotificationDelivery> delivery = prepareDelivery(session, phase, executionId);
        registry.send(ownerId, WsEvent.of("session_list_update", sessionId, Map.of("phase", phase)));
        registry.sendWithResult(ownerId, WsEvent.of("session_status", sessionId, statusData))
                .whenCompleteAsync((result, error) -> {
                    if (delivery.isEmpty()) return;
                    boolean delivered = error == null && result != null && result.delivered();
                    try {
                        deliveryService.resolveWebSocket(delivery.get(), delivered);
                    } catch (Exception e) {
                        log.warn("Failed to resolve WS result for task notification: deliveryId={}, error={}",
                                delivery.get().getId(), e.getMessage());
                    }
                }, notificationExecutor);
    }

    private boolean isTerminalPhase(String phase) {
        return "COMPLETED".equals(phase) || "FAILED".equals(phase) || "CANCELLED".equals(phase);
    }

    private Optional<TaskNotificationDelivery> prepareDelivery(Session session, String phase, String executionId) {
        try {
            return deliveryService.prepare(session, phase, executionId);
        } catch (Exception e) {
            log.warn("Failed to prepare task notification delivery: sessionId={}, error={}",
                    session.getId(), e.getMessage());
            return Optional.empty();
        }
    }
}

File: backend-ts/src/session/task-terminal.service.ts
import type { Session } from './types.js';
import type { SessionService } from './session.service.js';
import type { StreamingWsRegistry } from './ws/streaming-ws-registry.js';
import { delivered } from './ws/streaming-ws-registry.js';
import { wsEvent } from './ws/ws-event.js';
import type { SessionTreeSignalPublisher } from '../harness/approval/session-tree-signal-publisher.js';
import type { TaskNotificationDeliveryService } from '../notification/task/delivery.service.js';
import type { TaskNotificationDelivery } from '../notification/task/types.js';
import { WEIXIN_PROJECT_KEY } from '../domain/types.js';

const TERMINAL = new Set(['COMPLETED', 'FAILED', 'CANCELLED']);

export class TaskTerminalService {
  constructor(
    private readonly sessionService: SessionService,
    private readonly registry: StreamingWsRegistry,
    private readonly deliveryService: TaskNotificationDeliveryService,
    private readonly treeSignalPublisher: SessionTreeSignalPublisher,
    private readonly notificationExecutor: (fn: () => void | Promise<void>) => void = (fn) => {
      void Promise.resolve().then(fn);
    },
  ) {}

  async finishExecution(
    sessionId: number,
    userId: number | null | undefined,
    phase: string,
    executionId: string,
    failureReason?: string | null,
  ): Promise<void> {
    if (!TERMINAL.has(phase)) {
      throw new Error(`Unsupported terminal phase: ${phase}`);
    }
    const previous = await this.sessionService.getSession(sessionId);
    if (isTerminalPhase(previous.phase)) {
      console.info(
        `Ignoring terminal transition for already-terminal session: sessionId=${sessionId}, from=${previous.phase}, to=${phase}`,
      );
      return;
    }
    // FAILED 时将错误信息持久化到 runtimeStatusJson，刷新后前端可恢复
    if (phase === 'FAILED' && failureReason != null && failureReason.trim().length > 0) {
      await this.sessionService.updateRuntimeStatus(sessionId, { executionError: failureReason });
    } else {
      await this.sessionService.updateRuntimeStatus(sessionId, null);
    }
    await this.sessionService.updatePhase(sessionId, phase);
    await this.sessionService.markLastMessageFinished(sessionId);
    const session = await this.sessionService.getSession(sessionId);
    const ownerId = userId ?? session.userId;

    // 微信通道会话由定时任务等机器触发，终态不计未读，与 updatePhase 的 DB 写入保持一致
    const statusData: Record<string, unknown> = { phase, unread: session.projectKey !== WEIXIN_PROJECT_KEY };
    if (executionId != null && executionId.trim() !== '') {
      statusData.executionId = executionId;
    }

    const delivery = await this.prepareDelivery(session, phase, executionId, failureReason ?? null);
    if (ownerId != null) {
      this.registry.send(ownerId, wsEvent('session_list_update', sessionId, { phase }));
      void this.registry.sendWithResult(ownerId, wsEvent('session_status', sessionId, statusData)).then((result) => {
        if (delivery == null) return;
        this.notificationExecutor(async () => {
          try {
            await this.deliveryService.resolveWebSocket(delivery, delivered(result));
          } catch (e) {
            console.warn(
              `Failed to resolve WS result for task notification: deliveryId=${delivery.id}, error=${(e as Error).message}`,
            );
          }
        });
      });
    }

    if (session.sessionType === 'SIDE_TASK' && session.parentSessionId != null) {
      this.treeSignalPublisher.publish(session.parentSessionId);
    } else if (session.sessionType !== 'SUBAGENT') {
      // 主任务自身进入终态时也要重算并下发 treeRunning，否则前端列表里的
      // treeRunning 会停留在旧值（true），导致蓝色“执行中”圆点不转绿。
      this.treeSignalPublisher.publish(sessionId);
    }
  }

  private async prepareDelivery(
    session: Session,
    phase: string,
    executionId: string,
    failureReason: string | null,
  ): Promise<TaskNotificationDelivery | null> {
    try {
      return await this.deliveryService.prepare(session as never, phase, executionId, failureReason);
    } catch (e) {
      console.warn(`Failed to prepare task notification delivery: sessionId=${session.id}, error=${(e as Error).message}`);
      return null;
    }
  }
}

function isTerminalPhase(phase: string | null | undefined): boolean {
  return phase === 'COMPLETED' || phase === 'FAILED' || phase === 'CANCELLED';
}
